src/Compilers/Core/Portable/Diagnostic/DiagnosticInfo.cs

```diff
@@ -122,6 +122,25 @@ internal static void AssertMessageSerializable(object[] args)
             }
         }
 
+        [Conditional("DEBUG")]
+        internal static void AssertExpectedMessageArgumentsLength(CommonMessageProvider messageProvider, int errorCode, int actualLength)
+        {
+#if DEBUG
+            if (errorCode <= 0)
+            {
+                return;
+            }
+            string message = messageProvider.LoadMessage(errorCode, language: null) ?? string.Empty;
+            var matches = Regex.Matches(message, @"\{\d+}");
+            int expectedLength = 0;
+            foreach (Match match in matches)
+            {
+                expectedLength = Math.Max(int.Parse(match.Value[1..^1]) + 1, expectedLength);
+            }
+            Debug.Assert(expectedLength == actualLength);
```

Comment:
Would `Debug.Assert(expectedLength == matches.Count)` be valid (in addition to the existing assert)?

Reply:
`matches` has entries for each argument use, even duplicates, so `matches.Count` may not equal `expectedLength`. But it makes sense to check that all arguments are used, so I've used a `BitVector` instead, thanks.

There is one resource string that fails this new check: `ERR_IdentifierExpectedKW`. I've disabled the check for that `ErrorCode`, rather than changing the resource string and the arguments array, to avoid breaking translated resources. The assert is useful for new cases where arguments are unused though.